web/api/v1/api.go

```diff
@@ -553,6 +609,53 @@ func (api *API) labelValues(r *http.Request) (result apiFuncResult) {
 	return apiFuncResult{vals, nil, warnings, closer}
 }
 
+// LabelValuesByMatchers uses matchers to filter out matching series, then label values are extracted.
+func labelValuesByMatchers(sets []storage.SeriesSet, name string) ([]string, storage.Warnings, error) {
+	set := storage.NewMergeSeriesSet(sets, storage.ChainedSeriesMerge)
+	labelValuesSet := make(map[string]struct{})
+	for set.Next() {
+		series := set.At()
+		labelValue := series.Labels().Get(name)
+		labelValuesSet[labelValue] = struct{}{}
+	}
+
+	warnings := set.Warnings()
+	if set.Err() != nil {
+		return nil, warnings, set.Err()
+	}
+	// Convert the map to an array.
+	labelValues := make([]string, 0, len(labelValuesSet))
+	for key := range labelValuesSet {
+		labelValues = append(labelValues, key)
+	}
+	sort.Strings(labelValues)
+	return labelValues, warnings, nil
+}
+
+// labelNamesByMatchers uses matchers to filter out matching series, then label names are extracted.
+func labelNamesByMatchers(sets []storage.SeriesSet) ([]string, storage.Warnings, error) {
+	set := storage.NewMergeSeriesSet(sets, storage.ChainedSeriesMerge)
```

Comment:
Would we be faster iterating over each set separately?

Reply:
This wasn't addressed.

Reply:
I updated it for the `LabelNames` part, do I need to do the same thing for `LabelValues`?

Reply:
It's more a question of which is more efficient? The chain will use more RAM anyway, and we should be consistent.

Reply:
Yes, you are right, using `MergeSeriesSet` uses more RAM.

![Screenshot from 2020-12-18 13-40-01](https://user-images.githubusercontent.com/25150124/102649065-8f1cc980-4136-11eb-8429-b1343c14d775.png)

Reply:
Updated, PTAL @brian-brazil